source/jetstream/iterator.go

```diff
@@ -0,0 +1,276 @@
+	subscription  *nats.Subscription
+}
+
+// IteratorParams contains incoming params for the NewIterator function.
+type IteratorParams struct {
+	Conn          *nats.Conn
+	BufferSize    int
+	Durable       string
+	Stream        string
+	Subject       string
+	SDKPosition   sdk.Position
+	DeliverPolicy nats.DeliverPolicy
+	AckPolicy     nats.AckPolicy
+}
+
+// NewIterator creates new instance of the Iterator.
+func NewIterator(ctx context.Context, params IteratorParams) (*Iterator, error) {
+	jetstream, err := params.Conn.JetStream()
+	if err != nil {
+		return nil, fmt.Errorf("get jetstream context: %w", err)
+	}
+
+	consumerConfig, err := getConsumerConfig(params)
+	if err != nil {
+		return nil, fmt.Errorf("get consumer config: %w", err)
+	}
+
+	consumerInfo, err := jetstream.AddConsumer(params.Stream, consumerConfig)
+	if err != nil {
+		return nil, fmt.Errorf("add jetstream consumer: %w", err)
+	}
+
+	messages := make(chan *nats.Msg, params.BufferSize)
+
+	subscription, err := jetstream.ChanSubscribe(params.Subject, messages,
+		nats.Durable(consumerInfo.Config.Durable),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("chan subscribe: %w", err)
+	}
+
+	return &Iterator{
+		conn:          params.Conn,
+		messages:      messages,
+		unackMessages: make([]*nats.Msg, 0),
+		jetstream:     jetstream,
+		consumerInfo:  consumerInfo,
+		subscription:  subscription,
+	}, nil
+}
+
+// HasNext checks is the iterator has messages.
+func (i *Iterator) HasNext(ctx context.Context) bool {
+	return len(i.messages) > 0
+}
+
+// Next returns the next record from the underlying messages channel.
+// It also appends messages to a unackMessages slice if the AckPolicy is not equal to AckNonePolicy.
+func (i *Iterator) Next(ctx context.Context) (sdk.Record, error) {
+	select {
+	case msg := <-i.messages:
+		sdkRecord, err := i.messageToRecord(msg)
+		if err != nil {
+			return sdk.Record{}, fmt.Errorf("convert message to record: %w", err)
+		}
+
+		if i.consumerInfo.Config.AckPolicy != nats.AckNonePolicy {
+			i.Lock()
+			i.unackMessages = append(i.unackMessages, msg)
+			i.Unlock()
+		}
+
+		return sdkRecord, nil
+
+	case <-ctx.Done():
+		return sdk.Record{}, ctx.Err()
+	}
+}
+
+// Ack acknowledges a message at the given position.
+func (i *Iterator) Ack(ctx context.Context, sdkPosition sdk.Position) error {
+	// if ack policy is 'none' just return nil here
+	if i.consumerInfo.Config.AckPolicy == nats.AckNonePolicy {
+		return nil
+	}
+
+	i.Lock()
+	defer i.Unlock()
+
+	if err := i.canAck(sdkPosition); err != nil {
+		return fmt.Errorf("message cannot be acknowledged: %w", err)
+	}
+
+	if err := i.unackMessages[0].Ack(); err != nil {
+		return fmt.Errorf("ack message: %w", err)
+	}
+
+	// remove acknowledged message from the slice
+	i.unackMessages = i.unackMessages[1:]
+
+	return nil
+}
+
+// Stop stops the Iterator, unsubscribes from a subject.
+func (i *Iterator) Stop() (err error) {
+	defer func() {
+		if r := recover(); r != nil {
```

Comment:
Is it can be crush when we do teardown?

Reply:
In the case when the `messages` channel is already closed. But yeah, it's merely impossible here. I deleted the recover

Reply:
Only if Conduit will call the `Stop` method twice, for some reason